### PR TITLE
Muxer

### PR DESCRIPTION
**Pull request by iwahbe**
Fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/830

A PR that shows consuming this API is https://github.com/pulumi/pulumi-aws/pull/2421/files#diff-34c57e622183cb0d8dd0d3f9eaa0861b3340120e9b2ad811bac7ac7be4cea4b1.

### CODE REVIEW

**Comment by t0yv0** [comment]
Could you `make lint` on this @iwahbe 

```
 Error: Error return value of `(github.com/pulumi/pulumi-terraform-bridge/x/muxer.Main).Server` is not checked (errcheck)
```

File: pkg/tfbridge/main.go [hunk 1]
```diff
@@ -67,7 +69,19 @@ func Main(pkg string, version string, prov ProviderInfo, pulumiSchema []byte) {
...
 	}
 }
+
+type Option func(*opts)
+
+type opts struct {
+	muxWith []func(*provider.HostClient) (pulumirpc.ResourceProviderServer, error)
+}
+
+func MuxWith(s func(*provider.HostClient) (pulumirpc.ResourceProviderServer, error)) Option {
```
Comment (t0yv0) [inline pkg/tfbridge/main.go:83]: Mark this experimental API in a comment please.

File: pkg/tfbridge/serve.go [hunk 2]
```diff
@@ -16,16 +16,52 @@ package tfbridge
...
+func Serve(module string, version string, info ProviderInfo, pulumiSchema []byte, options ...Option) error {
+	opts := opts{}
+	for _, applyOption := range options {
+		applyOption(&opts)
+	}
+
 	// Create a new resource provider server and listen for and serve incoming connections.
-	return provider.Main(module, func(host *provider.HostClient) (lumirpc.ResourceProviderServer, error) {
+	return provider.Main(module, func(host *provider.HostClient) (pulumirpc.ResourceProviderServer, error) {
+		if len(opts.muxWith) > 0 {
```
Comment (t0yv0) [inline pkg/tfbridge/serve.go:37]: Nit: the glue code here can go into x/muxer.

File: pkg/tfgen/generate.go [hunk 3]
```diff
@@ -850,9 +850,18 @@ func (g *Generator) Generate() error {
 		return errors.Wrapf(err, "failed to create Pulumi schema")
 	}
 
+	// Now push the schema through the rest of the generator.
+	return g.GenerateFromSchema(pulumiPackageSpec)
+}
+
+// GenerateFromSchema creates Pulumi packages from a pulumi schema and the information the
+// generator was initialized with.
+func (g *Generator) GenerateFromSchema(pulumiPackageSpec pschema.PackageSpec) error {
```
Comment (t0yv0) [inline pkg/tfgen/generate.go:859]: This new public API is reasonable. Separate PR for this kind of thing is nice for project history (nit).

File: pf/tfgen/main.go [hunk 4]
```diff
@@ -60,6 +70,171 @@ func Main(provider string, info tfbridge.ProviderInfo) {
...
 
+// Implements main() logic for a multi-provider build-time helper utility. By convention these utilities are
+// named pulumi-tfgen-$provider, for example when building a "random" provider the program would be called
+// pulumi-tfgen-random.
+//
+// The resulting binary is able to generate [Pulumi Package Schema] as well as provider SDK sources in various
+// programming languages supported by Pulumi such as TypeScript, Go, and Python.
+//
+// [Pulumi Package Schema]: https://www.pulumi.com/docs/guides/pulumi-packages/schema/
+func MainWithMuxer(provider string, infos ...tfbridge.Muxed) {
```
Comment (t0yv0) [inline pf/tfgen/main.go:81]: Mark this experimental please in the comment. Something in the API/package breakdown is a bit fishy, but not worth blocking the PR on.

File: pf/tfgen/main.go [hunk 5]
```diff
@@ -60,6 +70,171 @@ func Main(provider string, info tfbridge.ProviderInfo) {
...
+				fDefined := func(k string) bool {
+					_, defined := muxedInfo.DataSources[k]
+					if defined {
+						return true
+					}
+					_, defined = muxedInfo.ExtraFunctions[k]
+					return defined
+				}
+
+				for k, v := range info.Resources {
```
Comment (t0yv0) [inline pf/tfgen/main.go:153]: This code is mixing ProviderInfo structures?

File: pf/tfbridge/main.go [hunk 6]
```diff
@@ -71,8 +86,58 @@ func Main(ctx context.Context, pkg string, prov ProviderInfo, meta ProviderMetad
...
+	})
+	if err != nil {
+		cmdutil.ExitError(err.Error())
+	}
+}
+
+// A union of pf and sdk based ProviderInfo for use in MainWithMuxer.
+//
+// Exactly 1 field of this struct should hold a value
+type Muxed struct {
```
Comment (t0yv0) [inline pf/tfbridge/main.go:133]: Mark this experimental please in the comment.

File: pf/tfbridge/main.go [hunk 7]
```diff
@@ -71,8 +86,58 @@ func Main(ctx context.Context, pkg string, prov ProviderInfo, meta ProviderMetad
 		fmt.Println(version)
 		os.Exit(0)
 	}
+}
 
-	if err := serve(ctx, pkg, prov, meta); err != nil {
+// Implements main() or a bridged Pulumi plugin, complete with argument parsing.
+func MainWithMuxer(ctx context.Context, pkg string, meta ProviderMetadata, infos ...Muxed) {
```
Comment (t0yv0) [inline pf/tfbridge/main.go:92]: Also experimental in the comment.

File: x/muxer/mapping.go [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+type ComputedMapping struct {
+	// mapping is an implementation detail.
+	//
+	// Right now, we want to expose the same information to the user, but that doesn't
+	// need to be true forever.
+	mapping
+}
+
+func Mapping(schemas []schema.PackageSpec) (ComputedMapping, schema.PackageSpec, error) {
+	// TODO Insert sanity checks and return an error on conflicts.  For example right
```
Comment (t0yv0) [inline x/muxer/mapping.go:33]: Very much worth doing especially for basic things like both providers define the same resource. I think we have to do it. Please let's link a ticket and let's do it soon, I can help perhaps.

File: x/muxer/muxer.go [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+		return errs[0]
+	default:
+		return (&multierror.Error{Errors: errs}).ErrorOrNil()
+	}
+}
+
+// Call n similar functions, returning a slice of their results.
+//
+// It is safe to assume that each function lines up with it's result.
+func asyncJoin[T any](f []func() T) []T {
```
Comment (t0yv0) [inline x/muxer/muxer.go:446]: Much fun here.

File: x/muxer/main.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+//     resource and function requests.
+//
+//   - Attach: `Attach` is never called on Muxed providers. Instead the host passed into
+//     `Mux` is replaced. If subsidiary servers where constructed with the same `host` as
+//     passed to `Mux`, then they will observe the new `host` spurred by `Attach`.
+//
+//   - GetMapping: TODO. GetMapping returns an opaque blob. It's not clear how to merge
+//     blobs. We might require that only one provider responds with a mapping, and just
+//     forward the first provider that responds.
+type Main struct {
```
Comment (t0yv0) [inline x/muxer/main.go:87]: This is not a great name for the API. Since we're in experimental its' ok we'll fix it up later.

**Comment by t0yv0** [comment]
Thanks for ramping this up quickly @iwahbe ! This is becoming a critical need for us and thank you for stepping in here. Excellent progress in short amount of time.

My overall sense is that this should go in once we have confirmation that a basic end to end use can such as AWS is working, preferably with an integration test covering the extra resources coming from AWS or PF.

The API naming and package partition strike me as a bit confused, the will need follow up work to get right so its' good to have all he added APIs marked experimental.

There is a lot of complexity in DiffConfig/Configure that AFAIK we have no urgent need for, therefore I'm glossing over reviewing those bits. It's nice to think about but we can deal with that problem when we have an actual provider with diverging configurations. Which can also inform the design, perhaps this is a kind of problem that a provider author will need to configure with more control rather than relying on a automating schema mixing, perhaps schema mixing will be useful, it's hard to anticipate.

Now the basic functionality of dispatching resources and function operations needs simple and rock solid and handle exceptions etc. I very much would like the TODOs around error handling here ticketed and done shortly after this PR lands.

The mixing of MarshallableProvider info response to the Schema needs a ticket linked to the TODO as well, we need that for tf2pulumi to work with muxed providers.

File: pkg/tfbridge/serve.go (hunk 2 above)
Comment (iwahbe) [reply pkg/tfbridge/serve.go:37]: Can you explain how? The glue code here calls bridge functions like `metadata.Get` and `NewProvider`. Right now the muxer does not depend on the rest of the bridge, and I would like to keep it that way.

File: pf/tfgen/main.go (hunk 5 above)
Comment (iwahbe) [reply pf/tfgen/main.go:153]: Yes. This is how we recover standard "unmapped resource/datasource" error messages.

File: x/muxer/mapping.go (hunk 8 above)
Comment (iwahbe) [reply x/muxer/mapping.go:33]: https://github.com/pulumi/pulumi-terraform-bridge/issues/949

**Review by t0yv0 — APPROVED** [review]

File: pf/tests/testdata/updateprogram/Pulumi.yaml [hunk 11]
```diff
@@ -17,6 +17,17 @@ resources:
...
       setOptionals: [ "${r.requiredInputStringCopy}", a, b ]
+
+  randomNumber:
+    type: muxedrandom:RandomInteger
+    properties:
+      min: 2
+      max: 2
+
+  pseudoRandomNumber:
+    type: muxedrandom:RandomHumanNumber
```
Comment (iwahbe) [inline pf/tests/testdata/updateprogram/Pulumi.yaml:28]: `pseudoRandomNumber` is a SDKv2 defined resource, while `randomInteger` is a PF defined resource. 

This gives us a basic end-to-end integration test for the muxer.

CC @t0yv0

Comment (t0yv0) [reply pf/tests/testdata/updateprogram/Pulumi.yaml:28]: Brilliant. Thank you!

**Pull request merged**